### PR TITLE
Wrap afw::display with pybind11

### PR DESCRIPTION
**Pull request by pschella**
This includes the following tests:

    - testImage.py
    - testOffsetImage.py
    - testStatistics.py
    - testImagePca.py
    - testMaskedImageIO.py
    - testWcs1.py
    - testStacker.py
    - testMask.py
    - testApproximate.py
    - testDisplay.py
    - testConvolve.py
    - test1079.py
    - testMaskedImage.py
    - testBackground
    - testWarpExposure.py

### CODE REVIEW

File: tests/testBackground.py [hunk 1]
```diff
@@ -735,7 +735,7 @@ def testApproximate(self):
 
             idiff = interpImage.getArray() - interpImage2.getArray()
             adiff = approxImage.getArray() - approxImage2.getArray()
-
+
```
Comment (fred3m) [inline tests/testBackground.py]: please remove extra whitespace

File: tests/testBackground.py [hunk 2]
```diff
@@ -672,13 +672,13 @@ def assertBackgroundList(bgl):
...
-        for i, j in zip(new, backgroundList):
-            self.assertBackgroundEqual(i[0], j[0])
-            self.assertEqual(i[1:], j[1:])
+#        # Check pickling
+#        new = pickle.loads(pickle.dumps(backgroundList))
+#        assertBackgroundList(new)
+#        self.assertEqual(len(new), len(backgroundList))
+#        for i, j in zip(new, backgroundList):
+#            self.assertBackgroundEqual(i[0], j[0])
+#            self.assertEqual(i[1:], j[1:])
```
Comment (fred3m) [inline tests/testBackground.py]: Make sure to update DM-8415

File: tests/testMask.py [hunk 3]
```diff
@@ -349,7 +349,7 @@ def testInterpret(self):
             self.assertEqual(im.interpret(bitmask), p)
             im.getArray()[0, i] = bitmask
             self.assertEqual(im.getAsString(i, 0), p)
-        self.assertEqual(self.Mask.interpret(allBits), ",".join(planes.keys()))
+        self.assertEqual(self.Mask.interpret(allBits), ",".join(sorted(planes.keys())))
```
Comment (fred3m) [inline tests/testMask.py]: I remember a discussion about this on slack but I can't recall the final conclusion. I know that planes is supposed to be a dict with sorted keys, can you remind me (and future developers) why it is ok that planes is not sorted anymore?

File: tests/testStacker.py [hunk 4]
```diff
@@ -133,20 +133,20 @@ def testRequestMoreThanOneStat(self):
...
-        imgList = afwImage.vectorImageF()
+        imgList = []
         for val in self.values:
             img = afwImage.ImageF(afwGeom.Extent2I(self.nX, self.nY), val)
-            imgList.push_back(img)
+            imgList.append(img)
 
         def tst():
-            afwMath.statisticsStack(imgList, afwMath.MEAN | afwMath.MEANCLIP, sctrl)
+            afwMath.statisticsStack(imgList, afwMath.Property(afwMath.MEAN | afwMath.MEANCLIP), sctrl)
```
Comment (fred3m) [inline tests/testStacker.py]: Why is this needed? More importantly, are there other places downstream that used afwMath.MEAN and afwMath.MEANCLIP?

File: python/lsst/afw/image/image.cc [hunk 5]
```diff
@@ -224,22 +296,35 @@ PYBIND11_PLUGIN(_image) {
         .value("LOCAL", ImageOrigin::LOCAL)
         .export_values();
 
+    declareImageBase<int>(mod, "I");
     declareImageBase<float>(mod, "F");
     declareImageBase<double>(mod, "D");
-    declareImageBase<int>(mod, "I");
```
Comment (fred3m) [inline python/lsst/afw/image/image.cc]: Should F and D be switched as well?

File: python/lsst/afw/geom/box.py [hunk 6]
```diff
@@ -67,3 +67,5 @@ def _Box2D_getCorners(self):
 del _Box2D_getSlices
 del _Box2D_getCorners
 
+BoxI = Box2I
+BoxD = Box2D
```
Comment (pschella) [inline python/lsst/afw/geom/box.py]: Yes

File: python/lsst/afw/image/image.cc [hunk 7]
```diff
@@ -211,6 +265,24 @@ void addImageSliceOperators(py::class_<Image<PixelT>,
...
+	cls.def("convertI", [](Image<PixelT> const & self) { return Image<int>(self, true); });
+	cls.def("convertF", [](Image<PixelT> const & self) { return Image<float>(self, true); });
+	cls.def("convertD", [](Image<PixelT> const & self) { return Image<double>(self, true); });
+	cls.def("convertU", [](Image<PixelT> const & self) { return Image<std::uint16_t>(self, true); });
+	cls.def("convertL", [](Image<PixelT> const & self) { return Image<std::uint64_t>(self, true); });
+
+	cls.def("convertFloat", [](Image<PixelT> const & self) { return Image<float>(self, true); });
+	cls.def("convertDouble", [](Image<PixelT> const & self) { return Image<double>(self, true); });
+}
+
```
Comment (pschella) [inline python/lsst/afw/image/image.cc]: Same

File: python/lsst/afw/image/image.cc (hunk 5 above)
Comment (pschella) [inline python/lsst/afw/image/image.cc]: Hmm, I always just used I, F, D, U, L. For this case it really shouldn't matter. But for overloads it might.

File: python/lsst/afw/image/image.cc [hunk 8]
```diff
@@ -224,22 +296,35 @@ PYBIND11_PLUGIN(_image) {
...
     declareImageBase<std::uint64_t>(mod, "L");
 
+    auto clsImageI = declareImage<int>(mod, "I");
     auto clsImageF = declareImage<float>(mod, "F");
     auto clsImageD = declareImage<double>(mod, "D");
-    declareImage<int>(mod, "I");
-    declareImage<std::uint16_t>(mod, "U");
-    declareImage<std::uint64_t>(mod, "L");
+    auto clsImageU = declareImage<std::uint16_t>(mod, "U");
+    auto clsImageL = declareImage<std::uint64_t>(mod, "L");
```
Comment (pschella) [inline python/lsst/afw/image/image.cc]: Again, for defining types it really doesn't matter. But for overloads it does and the order might be wrong. But we need to look at this on a case by case basis.

File: python/lsst/afw/image/mask.cc [hunk 9]
```diff
@@ -61,16 +61,21 @@ void declareMask(py::module & mod, std::string const & suffix) {
...
-    cls.def("__ixor__", (Mask<MaskPixelT>& (Mask<MaskPixelT>::*)(MaskPixelT const)) &Mask<MaskPixelT>::operator^=);
+    cls.def("__ior__", [](Mask<MaskPixelT> & self, Mask<MaskPixelT> & other) { return self |= other; }, py::is_operator());
+    cls.def("__ior__", [](Mask<MaskPixelT> & self, MaskPixelT const other) { return self |= other; }, py::is_operator());
+    cls.def("__ior__", [](Mask<MaskPixelT> & self, int other) { return self |= other; }, py::is_operator());
+    cls.def("__iand__", [](Mask<MaskPixelT> & self, Mask<MaskPixelT> & other) { return self &= other; }, py::is_operator());
+    cls.def("__iand__", [](Mask<MaskPixelT> & self, MaskPixelT const other) { return self &= other; }, py::is_operator());
+    cls.def("__iand__", [](Mask<MaskPixelT> & self, int other) { return self &= other; }, py::is_operator());
+    cls.def("__ixor__", [](Mask<MaskPixelT> & self, Mask<MaskPixelT> & other) { return self ^= other; }, py::is_operator());
+    cls.def("__ixor__", [](Mask<MaskPixelT> & self, MaskPixelT const other) { return self ^= other; }, py::is_operator());
+    cls.def("__ixor__", [](Mask<MaskPixelT> & self, int other) { return self ^= other; }, py::is_operator());
```
Comment (pschella) [inline python/lsst/afw/image/mask.cc]: clang-format! :) I think I'll just run that anyway. The small differences with our style guide won't pop up in the pybind11 code so it shouldn't be a problem.

File: python/lsst/afw/image/mask.py [hunk 10]
```diff
@@ -1,7 +1,17 @@
...
 
 supportSlicing(MaskU)
 
+class Mask(with_metaclass(ABCMeta, object)):
+    pass
+
+for cls in (MaskU, ):
+    Mask.register(cls)
+    supportSlicing(cls)
+
```
Comment (pschella) [inline python/lsst/afw/image/mask.py]: No, the reason was to introduce the ABC `Mask` (in concert with `Image`, `MaskedImage` and `Exposure`) to be used for `isinstance` testing. Instead of using a regexp to test against the output of `__repr__` which is problematic and ugly.

File: python/lsst/afw/math/approximate.cc [hunk 11]
```diff
@@ -22,24 +22,64 @@
...
+    py::class_<Class, std::shared_ptr<Class>> cls(mod, ("Approximate" + suffix).c_str());
+
+    cls.def("getImage", &Class::getImage,
+            "orderX"_a=-1, "orderY"_a=-1);
+    cls.def("getMaskedImage", &Class::getMaskedImage,
+            "orderX"_a=-1, "orderY"_a=-1);
+
+
+    mod.def("makeApproximate", (PTR(Approximate<PixelT>) (*)(std::vector<double> const &, std::vector<double> const &, image::MaskedImage<PixelT> const&, geom::Box2I const&, ApproximateControl const&)) makeApproximate<PixelT>,
+            "x"_a, "y"_a, "im"_a, "bbox"_a, "ctrl"_a);
```
Comment (pschella) [inline python/lsst/afw/math/approximate.cc]: :)

File: python/lsst/afw/image/mask.cc (hunk 9 above)
Comment (fred3m) [inline python/lsst/afw/image/mask.cc]: That's fine with me

File: python/lsst/afw/math/background.cc [hunk 12]
```diff
@@ -22,24 +22,142 @@
...
+            "interpStyle"_a, "undersampleStyle"_a=THROW_EXCEPTION);
+    cls.def(("getImage" + suffix).c_str(), (PTR(lsst::afw::image::Image<PixelT>) (BackgroundMI::*)(std::string const &, std::string const &) const) &BackgroundMI::getImage<PixelT>,
+            "interpStyle"_a, "undersampleStyle"_a="THROW_EXCEPTION");
+    cls.def(("getImage" + suffix).c_str(), (PTR(lsst::afw::image::Image<PixelT>) (BackgroundMI::*)(lsst::afw::geom::Box2I const&, Interpolate::Style const, UndersampleStyle const) const) &BackgroundMI::getImage<PixelT>,
+            "bbox"_a, "interpStyle"_a, "undersampleStyle"_a=THROW_EXCEPTION);
+    cls.def(("getImage" + suffix).c_str(), (PTR(lsst::afw::image::Image<PixelT>) (BackgroundMI::*)(lsst::afw::geom::Box2I const&, std::string const &, std::string const &) const) &BackgroundMI::getImage<PixelT>,
+            "bbox"_a, "interpStyle"_a, "undersampleStyle"_a="THROW_EXCEPTION");
+    cls.def(("getImage" + suffix).c_str(), (PTR(lsst::afw::image::Image<PixelT>) (BackgroundMI::*)() const) &BackgroundMI::getImage<PixelT>);
+}
```
Comment (pschella) [inline python/lsst/afw/math/background.cc]: :)

File: python/lsst/afw/math/background.cc [hunk 13]
```diff
@@ -22,24 +22,142 @@
...
+    clsBackgroundControl.def("setApproximateControl", &BackgroundControl::setApproximateControl);
+    clsBackgroundControl.def("getApproximateControl", (PTR(ApproximateControl) (BackgroundControl::*)()) &BackgroundControl::getApproximateControl);
+
+    /* Note that, in this case, the holder type must be unique_ptr to enable usage
+     * of py::nodelete, which in turn is needed because Background has a protected
+     * destructor. Adding py::nodelete prevents pybind11 from calling the destructor
+     * when the pointer is destroyed. Thus care needs to be taken to prevent leaks.
+     * Basically Background should only ever be used as a base class (without data
+     * members). */
+    py::class_<Background, std::unique_ptr<Background, py::nodelete>> clsBackground(mod, "Background");
```
Comment (pschella) [inline python/lsst/afw/math/background.cc]: No, this is intended use.

File: python/lsst/afw/math/detail/convolve.cc [hunk 14]
```diff
@@ -30,11 +30,43 @@ namespace py = pybind11;
...
+
+namespace {
+template <typename OutImageT, typename InImageT>
+void declareByType(py::module & mod) {
+    mod.def("basicConvolve", (void (*)(OutImageT&, InImageT const&, lsst::afw::math::Kernel const&, lsst::afw::math::ConvolutionControl const&)) basicConvolve<OutImageT, InImageT>);
+    mod.def("basicConvolve", (void (*)(OutImageT&, InImageT const&, lsst::afw::math::DeltaFunctionKernel const&, lsst::afw::math::ConvolutionControl const&)) basicConvolve<OutImageT, InImageT>);
+    mod.def("basicConvolve", (void (*)(OutImageT&, InImageT const&, lsst::afw::math::LinearCombinationKernel const&, lsst::afw::math::ConvolutionControl const&)) basicConvolve<OutImageT, InImageT>);
+    mod.def("basicConvolve", (void (*)(OutImageT&, InImageT const&, lsst::afw::math::SeparableKernel const&, lsst::afw::math::ConvolutionControl const&)) basicConvolve<OutImageT, InImageT>);
+    mod.def("convolveWithBruteForce", (void (*)(OutImageT&, InImageT const&, lsst::afw::math::Kernel const&, lsst::afw::math::ConvolutionControl const&)) convolveWithBruteForce<OutImageT, InImageT>);
+}
```
Comment (pschella) [inline python/lsst/afw/math/detail/convolve.cc]: :)

File: python/lsst/afw/math/detail/convolve.cc [hunk 15]
```diff
@@ -30,11 +30,43 @@ namespace py = pybind11;
...
+template <typename PixelType1, typename PixelType2>
+void declareAll(py::module & mod) {
+    using M1 = image::MaskedImage<PixelType1, image::MaskPixel, image::VariancePixel>;
+    using M2 = image::MaskedImage<PixelType2, image::MaskPixel, image::VariancePixel>;
+
+    declareByType<image::Image<PixelType1>, image::Image<PixelType2>>(mod);
+    declareByType<M1, M2>(mod);
+}
+}
```
Comment (pschella) [inline python/lsst/afw/math/detail/convolve.cc]: done

File: python/lsst/afw/math/warpExposure.cc [hunk 16]
```diff
@@ -42,4 +124,5 @@ PYBIND11_PLUGIN(_warpExposure) {
     /* Members */
 
     return mod.ptr();
-}
\ No newline at end of file
+}
+}}} // lsst::afw::math
```
Comment (pschella) [inline python/lsst/afw/math/warpExposure.cc]: :(

File: tests/testBackground.py (hunk 2 above)
Comment (pschella) [inline tests/testBackground.py]: Will do.

File: tests/testBackground.py (hunk 1 above)
Comment (pschella) [inline tests/testBackground.py]: done

File: tests/testMask.py (hunk 3 above)
Comment (pschella) [inline tests/testMask.py]: I think the conclusion from that discussion was that they never were supposed to be sorted to begin with. And that we should fix errors as we encounter them.

File: tests/testStacker.py (hunk 4 above)
Comment (pschella) [inline tests/testStacker.py]: This is needed because `Property | Property` returns `int` and the function accepts a `Property`. There will be more places where this is needed. But there is no way around it. In pybind11 enums are just safer and sometimes require explicit conversions.

File: tests/testMask.py (hunk 3 above)
Comment (fred3m) [inline tests/testMask.py]: Just for the sake of keeping a record: the issue is that in C++ MaskPlanes are maps, which are essentially dictionaries with sorted keys. SWIG mapped this to a dict with sorted keys while pybind11 maps them to ordinary dicts. @RobertLuptonTheGood confirmed that MaskPlanes do not need to be sorted, so it is ok to just change the API so that it does not expect MaskPlane.keys() to be sorted.

File: include/lsst/afw/image/Image.h [hunk 17]
```diff
@@ -689,4 +689,4 @@ namespace image {
 
 }}}  // lsst::afw::image
 
-#endif
+#endif
```
Comment (PaulPrice) [inline include/lsst/afw/image/Image.h]: Need an EOL before EOF.
